### DISCUSSION TITLE
What's the shortest path to get SumatraPDF to read the content? i.e. Text-to-Speech support.

### DISCUSSION THREAD

**Discussion by enihcam** [sumatrapdfreader/sumatrapdf#3243] [Ideas]
There are many truncated sentences unsupported by various 3rd-party TTS tools.

![image](https://user-images.githubusercontent.com/7010097/215253139-39908457-b0d1-412a-a34a-da8bccbd37ab.png)

What's the shortest path to get SumatraPDF to read the content correctly?

If there's no feasible path, may I ask for Text-to-Speech support in SumatraPDF?

Thanks.

**Comment by enihcam** [comment]
Similar discussions:
https://github.com/sumatrapdfreader/sumatrapdf/discussions/2761#discussioncomment-3038457
https://github.com/sumatrapdfreader/sumatrapdf/issues/321#issuecomment-599857556

**Comment by GitHubRulesOK** [comment]
There are no concepts of words in an average PDF  so "truncated sentences unsupported by various 3rd-party TTS tools." does not apply
there are NEVER sentences in a PDF since each letter is standalone.

Unlike eBooks it was not designed for reading or extraction only placing bits of characters at a numeric position via look-up tables. Thus such abilities are often secondary bolt-on's to any PDF viewer application.

Auto Hyphenation is a complex topic itself since there are  many ways they can be generated as line terminations, like a carriage return which pdfs do not have (there are no wrapped lines) as two lines do not need to be stored sequentially, but can be each end of the file.

 consider if I wrote PDF wise (not real but very very close sudo code)

Lookup table
50,700 Part 1 He
50,700 Part 2 d
50,700 Part 3 l
50,700 Part 4 !
50,700 Part 5 Wor
50,700 Part 6 o

then in the page I put `1336 5324` it is easy to PRINT (or read aloud) as `Hello World!` it is not connected to any other line except by the look-up of same fonted table. (lines do not need to be sequential they are referenced by start of line position) note that each language can hyphenate those words differently.

to get a feel for how PDF readers need to handle PDF constructs by reversing see in both SumatraPDF and make a second copy as a notepad TXT file this pdf https://github.com/GitHubRulesOK/MyNotes/raw/master/demo.pdf 
I have not used single characters (life is too short) but to see the above method look how this line  behaves
`BT -1.20 Tc 0.00 Tw 999 TL /F3 1 Tf 15 000 000 15 200.000 420.000 Tm [(J)-1200(a)800(ne Austen)] TJ ET`

**Reply by enihcam** [reply]
No, probably we don't need to dive into that details. Let me clarify the scenario:
1. User opens a PDF file with SumatraPDF. [existing capability]
2. User select a word, a sentence or a paragraph in multi-lines. [existing capability]
3. SumatraPDF gets a copy of selected content, either automatically or manually.
4. Remove all the carriage returns in the copy buffer; if '-' is tailing, remove it as well.
5. Invoke OS built-in TTS API to read the content.

The first version might be ugly, but better than nothing.

**Comment by GitHubRulesOK** [comment]
SumatraPDF does not select more then a bunch of letters between 2 punctuations try a double click to see its limitation it is how many letters a-z are almost touching (whay you call a word)
Here I double click the "Watermark" group of letters then clipboard allows me to paste that in say the find box.
![image](https://user-images.githubusercontent.com/12754913/215305283-caa9fcec-77dc-4b6a-95da-98c65d933a2e.png)

 It is the user selects words or sentences or pages or paragraphs worth of ink by placing into the windows clipboard application THAT is where they need to be read or converted. Many 3rd parties can do that remove unexpected breaks but its highly subjective when you get to the nitty gritty of what language or font was used, is it en or em is it hard or soft as they all may be coded incorrectly by the PDF writer. Acrobat does a fairly good convert by using large teams of international programmers.

**Comment by akobashikawa** [comment]
Hi, I just found https://www.naturalreaders.com/online/ and drag and drop a pdf in it. It is very useful, I think. Using plus voices has a limited daily quota but choosing a free voice you can use without limited quota, apparently.

**Comment by GitHubRulesOK** [comment]
@akobashikawa 
Apparently that service can NOT read a PDF that works well by send from Sumatra PDF to other apps like Google translate.
It asks to use OCR when the text is clearly text in the PDF ! the core issue is that in a PDF the characters are binary that needs display as ASCII ANSI or UniCode and that then  uses a system of numeric replacements via imbedded fonts, that are not always seen at the surface hence copy and paste to text is often a problem if the numeric conversion is not understood. Here is has to use 32bit Unicode, Whereas 16 bit is the way PDF holds the characters.

![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/325226e7-5952-450e-8558-130076db1109)

**Reply by akobashikawa** [reply]
I am not sure understand the problem but perhaps you could try to open the pdf document with naturalreaders instead of import selected text.

**Reply by GitHubRulesOK** [reply]
@akobashikawa 
That was my point NaturalReaders **does not understand ALL PDF text** that was a test piece that worked well for another issue after ensuring it matched the Extended characters. The PDF was uploaded but was not recognised as textual by NaturalReaders, but SumatraPDF can send the text as text.

There are so many ways text can be coded inside a PDF. Which is one of the many reasons PDF is not designed as a reversibale format it is designed to display in inked format the encoded content, those letters we see are not the pages content but a rendering as if an extractable audio shape.

Neither Microsoft or Google can read that text aloud simply convert to text !! and Google translate is the one that thinks the PDF is scanned when clearly there are no images (perhaps Natural Readers is using Googles free API)


![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/1549c306-b421-422b-ab8e-8bda8b28920c)


`A` letter can be stored many ways as (A) or <01> or <0001> or (  ÿ) OR.... but all we see is the GLYPH `A` and it is only those that are conventional on convert from number to human letter that we can hear as letters or when streamed as letter groups such as a phonetic word.

As a PDF Acrobat can make those characters as accessible **Text** Exactly the same as SumatraPDF can do.
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/8e2d1047-3265-4cbd-8684-0e7166d4a85d)

This Audio reader can see the characters in the PDF no problem sending it there as PDF or blind text.

![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/39290b30-9d45-4525-8dfa-122e4f4174ec)
